docs/rfc/004-integer-type-casts.md

```diff
@@ -0,0 +1,231 @@
+e.g. to mask bits and achieve the same effect as
+the value-changing casts discussed below.
+
+From a point of view, this approach seems to match Leo's
+treatment of potentially erroneous situations like integer overflows:
+the principle is that developers should explicitly use
+operations that may overflow if that is their intention,
+rather than having those situation possibly occur unexpectedly.
+
+A value-preserving cast to a type
+whose range does not cover the original type's range
+implicitly expresses a developer expectation that the value
+is actually in the intersection of the two types' ranges,
+in the same way that the use of integer addition
+implicitly expresses the expectation that the addition does not overflow.
+
+Consider this somewhat abstract example:
+```
+... // some computations on u32 values, which could not be done with u16
+let r: u32 = ...; // this is the final result of the u32 operations above
+let s: u16 = r as u16; // but r is expected to fit in u16, so we cast it here
+```
+With value-preserving casts, the expectation mentioned above
+is checked by the Leo compiler during proof generation,
+in the same way as with integer overflow.
+
+In the example above,
+if instead the variable `s` is meant to contain the low 16 bits of `r`,
+e.g. in a cryptographic computation,
+then the value-preserving cast should be preceded by
+an explicit operation to obtain the low 16 bits, making the intent clear:
+```
+... // some computations on u32 values, which could not be done with u16
+let r: u32 = ...; // this is the final result of the u32 operations above
+let r_low16: u32 = r & 0xFFFF; // assuming we have bitwise ops and hex literals
+let s: u16 = r_low16 as u16; // no value change here
+```
+
+### Value-Changing Casts
+
+The second approach is the following:
+ 1. `uN` to `uM` with `N < M`: just change type of value.
+ 2. `uN` to `uM` with `N > M`: take low `M` bits of value.
+ 3. `iN` to `iM` with `N < M`: just change type of value.
+ 4. `iN` to `iM` with `N > M`: take low `M` bits of value.
+ 5. `uN` to `iM` with `N < M`: zero-extend to `M` bits and re-interpret as signed.
+ 6. `uN` to `iM` with `N > M`: take low `M` bits and re-interpret as signed.
+ 7. `uN` to `iN`: re-interpret as signed
+ 8. `iN` to `uM` with `N < M`: sign-extend to `M` bits and re-interpret as unsigned.
+ 9. `iN` to `uM` with `N > M`: take low `M` bits and re-interpret as unsigned.
+10. `iN` to `uN`: re-interpret as unsigned
+Except for the 1st and 3rd cases, the value may change.
+
+This approach is common in other programming languages.
+However, it should be noted that other programming languages
+typically do not check for overflow in integer operations either
+(at least, not for production code).
+Presumably, the behavior of type casts in those programming languages
+is motivated by efficiency of execution, at least in part.
+Since in Leo the input data is avaialable at compile time,
```

Comment:
Typo -> "available".   Also, do we want to distinguish between compile-to-IR-or-R1CS time and proof generation time in this discussion, or is that too tangential?

Reply:
Fixed, thanks.

Reply:
I think it may be too tangential too at this level. (If you agree, you or I can click the 'Resolve conversation' button here.)